float32   angle
int32   grid_x
int32   grid_y
